File: exercises/js/homework-nft.js
// ============================================================
// Homework Exercise: NFT Lifecycle
// Learn: How to mint, query, and burn NFTs on the XRP Ledger.
// You'll experience the full lifecycle of an NFToken.
// Reference materials:
// - https://learn.xrpl-commons.org/course/code-with-the-xrpl/lesson/mint-and-burn-nfts/
// - https://xrpl.org/docs/references/protocol/transactions/types/nftokenmint
// - https://xrpl.org/docs/references/protocol/transactions/types/nftokenburn
// - https://xrpl.org/docs/references/http-websocket-apis/public-api-methods/account-methods/account_nfts
// - https://js.xrpl.org/
// ============================================================

const xrpl = require('xrpl');
const fs = require('fs');
const { notifyProgress } = require('../../scripts/notify');

const TESTNET_URL = 'wss://s.altnet.rippletest.net:51233';

async function main() {
  const { wallets } = JSON.parse(fs.readFileSync('../../wallets.json', 'utf-8'));
  const wallet = xrpl.Wallet.fromSeed(wallets[0].seed);

  const client = new xrpl.Client(TESTNET_URL);
  await client.connect();
  console.log('Connected to XRPL Testnet');

  // ╔══════════════════════════════════════════════════════════════════╗
  // ║ TODO 1: Mint an NFT                                             ║
  // ║                                                                  ║
  // ║ HINT: {                                                          ║
  // ║   TransactionType: 'NFTokenMint',                                ║
  // ║   Account: wallet.address,                                       ║
  // ║   NFTokenTaxon: 0,                                               ║
  // ║   Flags: 8,                                                      ║
  // ║   TransferFee: 5000,                                             ║
  // ║   URI: xrpl.convertStringToHex('ipfs://YOUR_NAME_HERE')          ║
  // ║ }                                                                ║
  // ║                                                                  ║
  // ║ Submit: client.submitAndWait(mintTx, { wallet })                  ║
  // ╚══════════════════════════════════════════════════════════════════╝
  // YOUR CODE HERE

  console.log('NFT Minted!');

  // ╔══════════════════════════════════════════════════════════════════╗
  // ║ TODO 2: Get account NFTs and extract the NFTokenID              ║
  // ║                                                                  ║
  // ║ HINT: const nfts = await client.request({                        ║
  // ║         command: 'account_nfts',                                  ║
  // ║         account: wallet.address                                   ║
  // ║       });                                                         ║
  // ║       const tokenId = nfts.result.account_nfts[0].NFTokenID;      ║
  // ╚══════════════════════════════════════════════════════════════════╝
  // YOUR CODE HERE

  // ╔══════════════════════════════════════════════════════════════════╗
  // ║ TODO 3: Burn the NFT                                            ║
  // ║                                                                  ║
  // ║ HINT: {                                                          ║
  // ║   TransactionType: 'NFTokenBurn',                                ║
  // ║   Account: wallet.address,                                       ║
  // ║   NFTokenID: tokenId                                              ║
  // ║ }                                                                ║
  // ║                                                                  ║
  // ║ Submit: client.submitAndWait(burnTx, { wallet })                  ║
  // ╚══════════════════════════════════════════════════════════════════╝
  // YOUR CODE HERE

  console.log('NFT Burned!');

  // ╔══════════════════════════════════════════════════════════════════╗
  // ║ TODO 4: Query account_nfts again to verify it's gone            ║
  // ║                                                                  ║
  // ║ HINT: Same request as TODO 2, then log the count                 ║
  // ╚══════════════════════════════════════════════════════════════════╝
  // YOUR CODE HERE

  console.log(`\nExplorer: https://testnet.xrpl.org/accounts/${wallet.address}`);
  await notifyProgress('Homework: NFT Lifecycle');
  await client.disconnect();
}

main().catch(console.error);

File: exercises/python/homework_nft.py
"""
Homework Exercise: NFT Mint, Query, and Burn
=============================================
Mint an NFT, query it, then burn it on XRPL Testnet.

Reference materials:
- https://learn.xrpl-commons.org/course/code-with-the-xrpl/lesson/mint-and-burn-nfts/
- https://xrpl.org/docs/references/protocol/transactions/types/nftokenmint
- https://xrpl.org/docs/references/protocol/transactions/types/nftokenburn
- https://xrpl.org/docs/references/http-websocket-apis/public-api-methods/account-methods/account_nfts
- https://xrpl-py.readthedocs.io/en/stable/
"""

import json
from xrpl.clients import WebsocketClient
from xrpl.wallet import Wallet
from xrpl.models.transactions import NFTokenMint, NFTokenBurn
from xrpl.models.requests import AccountNFTs
from xrpl.transaction import submit_and_wait
from xrpl.utils import str_to_hex
import sys, os; sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..', 'scripts'))
from notify_progress import notify_progress

TESTNET_URL = 'wss://s.altnet.rippletest.net:51233'

# Load wallets
with open('../../wallets.json') as f:
    data = json.load(f)
    wallets = data['wallets']

wallet = Wallet.from_seed(wallets[0]['seed'])
print(f"Wallet: {wallet.address}")

with WebsocketClient(TESTNET_URL) as client:

    # ╔══════════════════════════════════════════════════════════════════╗
    # ║ TODO 1: Mint an NFT                                             ║
    # ║                                                                  ║
    # ║ HINT: NFTokenMint(account=wallet.address, nftoken_taxon=0,       ║
    # ║       flags=8, transfer_fee=5000,                                ║
    # ║       uri=str_to_hex('ipfs://YOUR_NAME_HERE'))                   ║
    # ╚══════════════════════════════════════════════════════════════════╝
    mint_tx = None  # Replace with NFTokenMint(...)
    # mint_result = submit_and_wait(mint_tx, client, wallet)

    # ╔══════════════════════════════════════════════════════════════════╗
    # ║ TODO 2: Query NFTs to get the nftoken_id                        ║
    # ║                                                                  ║
    # ║ HINT: nfts_response = client.request(                            ║
    # ║           AccountNFTs(account=wallet.address))                    ║
    # ║       nftoken_id = nfts_response.result['account_nfts'][-1]      ║
    # ║                    ['NFTokenID']                                  ║
    # ╚══════════════════════════════════════════════════════════════════╝
    nftoken_id = None  # Replace with query logic

    # ╔══════════════════════════════════════════════════════════════════╗
    # ║ TODO 3: Burn the NFT                                            ║
    # ║                                                                  ║
    # ║ HINT: NFTokenBurn(account=wallet.address,                        ║
    # ║                   nftoken_id=nftoken_id)                         ║
    # ╚══════════════════════════════════════════════════════════════════╝
    burn_tx = None  # Replace with NFTokenBurn(...)
    # burn_result = submit_and_wait(burn_tx, client, wallet)

    # ╔══════════════════════════════════════════════════════════════════╗
    # ║ TODO 4: Query NFTs again to verify the NFT is gone              ║
    # ║                                                                  ║
    # ║ HINT: Same as TODO 2 - check the count                          ║
    # ╚══════════════════════════════════════════════════════════════════╝
    pass

# Explorer: https://testnet.xrpl.org/accounts/{address}/nfts
print(f"\nExplorer: https://testnet.xrpl.org/accounts/{wallet.address}/nfts")
notify_progress('Homework: NFT Lifecycle')
